api/v1beta1/grpc_disruption.go

```diff
@@ -0,0 +1,173 @@
+	// +ddmark:validation:Enum=OK;CANCELED;UNKNOWN;INVALID_ARGUMENT;DEADLINE_EXCEEDED;NOT_FOUND;ALREADY_EXISTS;PERMISSION_DENIED;RESOURCE_EXHAUSTED;FAILED_PRECONDITION;ABORTED;OUT_OF_RANGE;UNIMPLEMENTED;INTERNAL;UNAVAILABLE;DATA_LOSS;UNAUTHENTICATED
+	ErrorToReturn string `json:"error,omitempty"`
+	// +kubebuilder:validation:Enum={}
+	// +ddmark:validation:Enum={}
+	OverrideToReturn string `json:"override,omitempty"`
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:validation:Maximum=100
+	// +ddmark:validation:Minimum=1
+	// +ddmark:validation:Maximum=100
+	QueryPercent int `json:"query_pct,omitempty"`
+}
+
+// Validate validates that there are no missing hostnames or records for the given grpc disruption spec
+func (s GRPCDisruptionSpec) Validate() error {
+	if s.Port == 0 {
+		return errors.New("the gRPC disruption does not specify the port which the target service exposes, but port must be specified")
+	}
+
+	if len(s.Endpoints) == 0 {
+		return errors.New("the gRPC disruption was selected with no endpoints specified, but endpoints must be specified")
+	}
+
+	queryPctByEndpoint := make(map[string]int)
+	unquantifiedAlts := make(map[string]int)
+
+	for _, alteration := range s.Endpoints {
+		if alteration.TargetEndpoint == "" {
+			return errors.New("some list items in gRPC disruption are missing endpoints; specify a TargetEndpoint for each item in the list")
+		}
+
+		// check that endpoint is not already configured such that the sum of mangled queryPercents total to more than 100%
+		if alteration.QueryPercent == 0 {
+			if count, ok := unquantifiedAlts[alteration.TargetEndpoint]; ok {
+				unquantifiedAlts[alteration.TargetEndpoint] = count + 1
+
+				pctClaimed := 100 - queryPctByEndpoint[alteration.TargetEndpoint]
+				if pctClaimed < count+1 {
+					return fmt.Errorf("%s will never return some alterations because alterations exceed 100%% of possible queries", alteration.TargetEndpoint)
+				}
+			} else {
+				unquantifiedAlts[alteration.TargetEndpoint] = 1
+			}
+		} else {
+			if totalQueryPercent, ok := queryPctByEndpoint[alteration.TargetEndpoint]; ok {
+				if alteration.QueryPercent > 0 {
+					queryPctByEndpoint[alteration.TargetEndpoint] = totalQueryPercent + alteration.QueryPercent
+					if queryPctByEndpoint[alteration.TargetEndpoint] > 100 {
+						return errors.New("total queryPercent of all alterations applied to endpoint %s is over 100%; modify them to so their total is 100% or less")
+					}
+				}
+			} else {
+				queryPctByEndpoint[alteration.TargetEndpoint] = alteration.QueryPercent
+			}
+		}
+
+		// check that exactly one of ErrorToReturn or OverrideToReturn is configured
+		if alteration.ErrorToReturn != "" {
+			if alteration.OverrideToReturn == "" {
+				if _, ok := ErrorMap[alteration.ErrorToReturn]; !ok {
```

Comment:
Context here:

* Originally, I felt good about not having `ErrorMap` here because I validate `gRPC` strings.
* However, I do not think we unit-test `ddmark`.
* In order to make sure that the errors being request will actually work on the server side, we need to make sure the server is using string mappings which work with `ddmark`. And in fact if they're not in sync, then `chaos-controller` should change `ddmark` rather than have our users update their DisruptionListener services (for example with `CANCELED` vs `CANCELLED`.
* By having the `ErrorMap` mapping here and importing it in the server code, we can unit test that the mapping is the same, meaning the only thing that can go wrong is the ddmark.

Reply:
We could add unit testing for the ddmark inside this test https://github.com/DataDog/chaos-controller/pull/396/files#diff-808195d7ed62fe71cd613f1261479ea92b280d2026e5a44e307a7f79ac58b4e9R182 ? That unit test could pull directly from ErrorMap, accomplishing the same goal but without validating twice?

Reply:
Need to figure out what file path is, my test fails for this

Reply:
updated!